### PR TITLE
MC-19366: Adds GraphQL sniffs

### PR DESCRIPTION
**Pull request by jean-bernard-valentaten**
Sniffing GraphQL schemas was added. The following rules/sniffs were added:
* `enum`, `type` and `interface` names must be `UpperCamelCase`
* Field names of `type` and `interface`  must be `snake_case`
* Argument names of fields must be `camelCase`
* `query` and `mutation` names must be `camelCase`

In order to be able to tokenize a GraphQL file, we now make use of the lexer bundled in the package `webonyx/graphql-php`.

### CODE REVIEW

File: Magento2/Tests/GraphQL/ValidTypeNameUnitTest.graphqls [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+type ValidCamelCaseTypeWith1Number {}
+type ValidCamelCaseTypeWith12345Numbers {}
+type 5InvalidCamelCaseTypeStartingWithNumber {}
+type ValidCamelCaseTypeEndingWithNumber4 {}
+interface ValidCamelCaseInterfaceWith1Number {}
+interface ValidCamelCaseInterfaceWith12345Numbers {}
+interface 5InvalidCamelCaseInterfaceStartingWithNumber {}
+interface ValidCamelCaseInterfaceEndingWithNumber4 {}
+enum ValidCamelCaseEnumWith1Number {}
+enum ValidCamelCaseEnumWith12345Numbers {}
```
Comment (melnikovi) [inline Magento2/Tests/GraphQL/ValidTypeNameUnitTest.graphqls]: Do we want to allow numbers?

File: Magento2/Tests/GraphQL/ValidFieldNameUnitTest.graphqls [hunk 2]
```diff
@@ -0,0 +1,31 @@
+type FooType {
+    # Valid snake case field names
+    valid_snake_case_field: String
```
Comment (paliarush) [inline Magento2/Tests/GraphQL/ValidFieldNameUnitTest.graphqls]: We may have to change this rule soon to enforce camel case as discussed on architecture meeting. Please comment out this rule for now (including the check for interfaces).

File: Magento2/Tests/GraphQL/ValidTypeNameUnitTest.graphqls (hunk 1 above)
Comment (paliarush) [reply Magento2/Tests/GraphQL/ValidTypeNameUnitTest.graphqls]: It is probably ok.

File: Magento2/Tests/GraphQL/ValidTypeNameUnitTest.graphqls [hunk 3]
```diff
@@ -0,0 +1,44 @@
+# Valid type names.
+type ValidCamelCaseType {}
+interface ValidCamelCaseInterface {}
+enum ValidCamelCaseEnum {}
```
Comment (paliarush) [inline Magento2/Tests/GraphQL/ValidTypeNameUnitTest.graphqls]: We need to validate enum values to be UPPERCASE.

**Comment by lenaorobei** [comment]
@jean-bernard-valentaten thanks, that is significant improvement!

Could you please demo it on public architecture meeting https://github.com/magento/architecture/issues/273?

I see that some rules are still under discussion https://github.com/magento/architecture/issues/269, so please be aware that we cannot merge it until final decision is made.

File: Magento2/Tests/GraphQL/ValidFieldNameUnitTest.graphqls (hunk 2 above)
Comment (jean-bernard-valentaten) [reply Magento2/Tests/GraphQL/ValidFieldNameUnitTest.graphqls]: Rule was commented out, see https://github.com/magento/magento-coding-standard/pull/141/commits/2a8d52183bb612c3e4b520fb9b6d8a664932d9a9

**Comment by jean-bernard-valentaten** [comment]
@lenaorobei Thank you for the note, I fixed that issue and a couple of follow up issues

File: Magento2/Sniffs/GraphQL/AbstractGraphQLSniff.php [hunk 4]
```diff
@@ -0,0 +1,76 @@
+<?php
+/**
```
Comment (melnikovi) [inline Magento2/Sniffs/GraphQL/AbstractGraphQLSniff.php]: Need to have Magento copyright in all files.

File: PHP_CodeSniffer/Tokenizers/GraphQL.php [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+{
+
+    /**
+     * Defines how GraphQL token types are mapped to PHP token types.
+     *
+     * @var array
+     */
+    private $tokenTypeMap = [
+        Token::AT           => 'T_DOC_COMMENT_TAG',
+        Token::BANG         => null, //TODO Should we map this to a specific type
```
Comment (melnikovi) [inline PHP_CodeSniffer/Tokenizers/GraphQL.php]: Need to finalize the mapping.

File: PHP_CodeSniffer/Tokenizers/GraphQL.php [hunk 6]
```diff
@@ -0,0 +1,258 @@
+<?php
+/**
+ * Copyright © Magento. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+namespace PHP_CodeSniffer\Tokenizers;
```
Comment (melnikovi) [inline PHP_CodeSniffer/Tokenizers/GraphQL.php]: Should be under Magento namespace.

Comment (jean-bernard-valentaten) [reply PHP_CodeSniffer/Tokenizers/GraphQL.php]: Agreed, unfortunately PHP_CodeSniffer is not very fond of additional tokenizers and have therefore hardcoded the namespace under which tokenizers reside to `PHP_CodeSniffer\Tokenizers`: https://github.com/squizlabs/PHP_CodeSniffer/blob/master/src/Files/File.php#L562

File: PHP_CodeSniffer/Tokenizers/GraphQL.php (hunk 5 above)
Comment (jean-bernard-valentaten) [reply PHP_CodeSniffer/Tokenizers/GraphQL.php]: Fixed in https://github.com/magento/magento-coding-standard/pull/141/commits/43ef0a4b8d14197f8ec34cc97161a45a6d71b45a

**Comment by lenaorobei** [comment]
I run against Magento 2 `app/code` and following false-positive findings were discovered:
```

FILE: app/code/Magento/CmsGraphQl/etc/schema.graphqls
----------------------------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
----------------------------------------------------------------------------------------------
 19 | WARNING | Argument name "String@doc(description" is not in CamelCase format
----------------------------------------------------------------------------------------------


FILE: app/code/Magento/CustomerGraphQl/etc/schema.graphqls
-------------------------------------------------------------------------------------------------------------------------------
FOUND 0 ERRORS AND 246 WARNINGS AFFECTING 246 LINES
-------------------------------------------------------------------------------------------------------------------------------
   7 | WARNING | Argument name "@doc(description" is not in CamelCase format
 138 | WARNING | Enum value "AF@doc(description:Afghanistan)" is not in SCREAMING_SNAKE_CASE format
 139 | WARNING | Enum value "AX@doc(description:Åland Islands)" is not in SCREAMING_SNAKE_CASE format
 140 | WARNING | Enum value "AL@doc(description:Albania)" is not in SCREAMING_SNAKE_CASE format
...
-------------------------------------------------------------------------------------------------------------------------------


FILE: app/code/Magento/DownloadableGraphQl/etc/schema.graphqls
----------------------------------------------------------------------------------------------------------------------------------------------
FOUND 0 ERRORS AND 2 WARNINGS AFFECTING 2 LINES
----------------------------------------------------------------------------------------------------------------------------------------------
 41 | WARNING | Enum value "FILE@deprecated(reason:`sample_url` serves to get the downloadable sample)" is not in SCREAMING_SNAKE_CASE format
 42 | WARNING | Enum value "URL@deprecated(reason:`sample_url` serves to get the downloadable sample)" is not in SCREAMING_SNAKE_CASE format
----------------------------------------------------------------------------------------------------------------------------------------------
```
Please add these cases to the unit tests as well.
Thanks.

File: Magento2/ruleset.xml [hunk 7]
```diff
@@ -522,6 +522,31 @@
     <rule ref="Squiz.WhiteSpace.SuperfluousWhitespace.EmptyLines">
         <severity>0</severity>
     </rule>
+    <rule ref="Magento2.GraphQL.ValidTypeName">
```
Comment (lenaorobei) [inline Magento2/ruleset.xml]: Please respect rules order in the `ruleset` - by severity -> by name.

File: Magento2/ruleset.xml [hunk 8]
```diff
@@ -522,6 +522,31 @@
...
+        <severity>6</severity>
+        <type>warning</type>
+    </rule>
+    <rule ref="Magento2.GraphQL.ValidFieldName">
+        <!--
+         * Field name validation deactivated as per request
+         * https://github.com/magento/magento-coding-standard/pull/141#discussion_r323349939
+        -->
+        <exclude name="Magento2.GraphQL.ValidFieldName" />
+        <severity>6</severity>
```
Comment (lenaorobei) [inline Magento2/ruleset.xml]: If this rule needs to be deactivated please set the severity to 0.

File: composer.json [hunk 9]
```diff
@@ -9,11 +9,17 @@
     "version": "4",
     "require": {
         "php": ">=5.6.0",
-        "squizlabs/php_codesniffer": "^3.4"
+        "squizlabs/php_codesniffer": "^3.4",
+        "webonyx/graphql-php": ">=0.12.6 <1.0"
```
Comment (lenaorobei) [inline composer.json]:
```suggestion
        "webonyx/graphql-php": "^0.12"
```

File: Magento2/ruleset.xml (hunk 7 above)
Comment (jean-bernard-valentaten) [reply Magento2/ruleset.xml]: Fixed in https://github.com/magento/magento-coding-standard/pull/141/commits/c539ed44c3b93284d9bc24e5e58c0271cbaa1d40

File: Magento2/ruleset.xml (hunk 8 above)
Comment (jean-bernard-valentaten) [reply Magento2/ruleset.xml]: Fixed in https://github.com/magento/magento-coding-standard/pull/141/commits/c539ed44c3b93284d9bc24e5e58c0271cbaa1d40

**Comment by jean-bernard-valentaten** [comment]
> I run against Magento 2 `app/code` and following false-positive findings were discovered:
> 
> ```
> 
> FILE: app/code/Magento/CmsGraphQl/etc/schema.graphqls
> ----------------------------------------------------------------------------------------------
> FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
> ----------------------------------------------------------------------------------------------
>  19 | WARNING | Argument name "String@doc(description" is not in CamelCase format
> ----------------------------------------------------------------------------------------------
> 
> 
> FILE: app/code/Magento/CustomerGraphQl/etc/schema.graphqls
> -------------------------------------------------------------------------------------------------------------------------------
> FOUND 0 ERRORS AND 246 WARNINGS AFFECTING 246 LINES
> -------------------------------------------------------------------------------------------------------------------------------
>    7 | WARNING | Argument name "@doc(description" is not in CamelCase format
>  138 | WARNING | Enum value "AF@doc(description:Afghanistan)" is not in SCREAMING_SNAKE_CASE format
>  139 | WARNING | Enum value "AX@doc(description:Åland Islands)" is not in SCREAMING_SNAKE_CASE format
>  140 | WARNING | Enum value "AL@doc(description:Albania)" is not in SCREAMING_SNAKE_CASE format
> ...
> -------------------------------------------------------------------------------------------------------------------------------
> 
> 
> FILE: app/code/Magento/DownloadableGraphQl/etc/schema.graphqls
> ----------------------------------------------------------------------------------------------------------------------------------------------
> FOUND 0 ERRORS AND 2 WARNINGS AFFECTING 2 LINES
> ----------------------------------------------------------------------------------------------------------------------------------------------
>  41 | WARNING | Enum value "FILE@deprecated(reason:`sample_url` serves to get the downloadable sample)" is not in SCREAMING_SNAKE_CASE format
>  42 | WARNING | Enum value "URL@deprecated(reason:`sample_url` serves to get the downloadable sample)" is not in SCREAMING_SNAKE_CASE format
> ----------------------------------------------------------------------------------------------------------------------------------------------
> ```
> 
> Please add these cases to the unit tests as well.
> Thanks.

Fixed in https://github.com/magento/magento-coding-standard/pull/141/commits/af9c04119912bb96ccd5e067c97772d7169f53b1 and https://github.com/magento/magento-coding-standard/pull/141/commits/c4069e0e44e3b16fc75c01df7c131d599ff8f67f

**Comment by lenaorobei** [comment]
Please also address that following message appears on `composer install` step: `Warning: The lock file is not up to date with the latest changes in composer.json. You may be getting outdated dependencies. Run update to update them.`

**Comment by jean-bernard-valentaten** [comment]
> Please also address that following message appears on `composer install` step: `Warning: The lock file is not up to date with the latest changes in composer.json. You may be getting outdated dependencies. Run update to update them.`

Fixed in https://github.com/magento/magento-coding-standard/pull/141/commits/4d99ee4797c2e881bd1e75bee78e5faeef2f2dd2

**Pull request merged**